Add dbg_stats_of()

mcostalba opened the pull request:
To compute the covariance of a set of
variables.

Usage is very simple. If you want to compute
covariance of parameters A, B and C, simply add
in sources, in the place you want to sample the values:

dbg_stats_of(A, B, C);

The only rule is that A, B and C should be convertible
to a double. There is no constrain on the numbers of parameters.

No functional change.

Stefano80 commented:
Hi Marco,

the skeleton is quite fleshy indeed. It works and seems to give meaningful results for me. My feedback:

0) A question for understanding: what does it  "typename..."  mean? I could not find the explanation on cppreference.com

1) Maybe it makes sense to change the definition in misc.h to

template<typename T, typename... Params>
void dbg_stats_of(T p, Params... parameters) {
  <Some form of dbg_mean_of>
  dbg_stats_add_sample(double(p));
  dbg_stats_of(parameters...);
}

Since you need the means anyway to compute the covariance (you are not subtracting E^2X yet) and so you would get also the mean in the stats

2) We have to decide whether we want variances+covariances or std+correlations. The first is simpler and you dont have to make divisions, so you also dont have problem with division by 0. The second gives an output which is much easier to understand for the users. I am neutral.

3) Maybe far fetched, but I wanted to give you the idea: why we dont just add a branch like the tuning branch with all this debug stuff? So we could get rid of it from main branch but we could maybe elaborate a bit more on what could be possible. (For instance, I just learned that it is possible to train regression models on streaming data quite easily, and so the need of dumping anything to an external file would be eradicated for ever).

mcostalba commented:
@Stefano80 

0) It is called variadic template, see:
https://en.wikipedia.org/wiki/Variadic_template
http://en.cppreference.com/w/cpp/language/parameter_pack

1) I have added the means and updated the PR

2) I'd go with the simplest first: there is always time to complicate it later :-)

3) This is a good idea! Actually even better is to add it to the tune branch itself. So that you have all in one place and using the 'tune' branch instead of master we are more free to indulge in added code lines :-)

Stefano80 commented:
0) Thank you, learned again something

1-2) The correct formula for line 187 I think is

( Covariances[i] - Sums[i % N]*Sums[i/N] / samplesCnt ) / samplesCnt 

since the covariance is E(XY) - EXEY. The formatting is very comfortable.

3) But maybe we should rename the branch to something like "dev"

Stefano80 commented:
Hi Marco, 
I made some more work on it, adding the samples count, the correct formula and some formatting. You find the changes in https://github.com/Stefano80/Stockfish/tree/mc_dbgStats

From my point of view the math and comfort is now good enough.

mcostalba commented:
@Stefano80 ok thanks, I will integrate into tune branch.

mcostalba commented:
I have added dbg_stats_of() to the [tune ](https://github.com/official-stockfish/Stockfish/tree/tune )branch.

Stefano80 commented:
Hi Marco, dbg_stats_of was somehow removed from tune branch

mcostalba commented:
@Stefano80 sorry, I have somehow messed the update. It should be fixed now.

Stefano80 commented:
Thanks Marco, just submitted a test based on that.

Lolligerhans commented:
@snicolet Oh it is already there, that is great! I have these questions:

1. Why are the "`Covariances`" left at the sum of products (i.e. Stefanos formula above is not applied at all)?
2. There is a function template, but any given type is always casted to `double. This makes we wonder: Why is the funciton templated?
